rpc/src/v1/impls/eth_pubsub.rs

```diff
@@ -0,0 +1,153 @@
+use jsonrpc_core::Error;
+use jsonrpc_macros::Trailing;
+use jsonrpc_macros::pubsub::{Sink, Subscriber};
+use jsonrpc_pubsub::SubscriptionId;
+
+use v1::helpers::{errors, Subscribers};
+use v1::metadata::Metadata;
+use v1::traits::EthPubSub;
+use v1::types::{pubsub, RichHeader};
+
+use ethcore::encoded;
+use ethcore::client::{BlockChainClient, ChainNotify, BlockId};
+use light::client::{LightChainClient, LightChainNotify};
+use parity_reactor::Remote;
+use util::{Mutex, H256, Bytes};
+
+/// Eth PubSub implementation.
+pub struct EthPubSubClient<C> {
+	handler: Arc<ChainNotificationHandler<C>>,
+	heads_subscribers: Arc<Mutex<Subscribers<Sink<pubsub::Result>>>>,
+}
+
+impl<C> EthPubSubClient<C> {
+	/// Creates new `EthPubSubClient`.
+	pub fn new(client: Arc<C>, remote: Remote) -> Self {
+		let heads_subscribers = Arc::new(Mutex::new(Subscribers::default()));
+		EthPubSubClient {
+			handler: Arc::new(ChainNotificationHandler {
+				client: client,
+				remote: remote,
+				heads_subscribers: heads_subscribers.clone(),
+			}),
+			heads_subscribers: heads_subscribers,
+		}
+	}
+
+	/// Returns a chain notification handler.
+	pub fn handler(&self) -> Arc<ChainNotificationHandler<C>> {
+		self.handler.clone()
+	}
+}
+
+/// PubSub Notification handler.
+pub struct ChainNotificationHandler<C> {
+	client: Arc<C>,
+	remote: Remote,
+	heads_subscribers: Arc<Mutex<Subscribers<Sink<pubsub::Result>>>>,
+}
+
+impl<C> ChainNotificationHandler<C> {
+	fn notify(&self, blocks: Vec<(encoded::Header, BTreeMap<String, String>)>) {
+		for subscriber in self.heads_subscribers.lock().values() {
+			for &(ref block, ref extra_info) in &blocks {
+				self.remote.spawn(subscriber
+					.notify(pubsub::Result::Header(RichHeader {
+						inner: block.into(),
+						extra_info: extra_info.clone(),
+					}))
+					.map(|_| ())
+					.map_err(|e| warn!(target: "rpc", "Unable to send notification: {}", e))
+				);
+			}
+		}
+	}
+}
+
+impl<C: LightChainClient> LightChainNotify for ChainNotificationHandler<C> {
+	fn new_headers(
+		&self,
+		headers: &[H256],
+	) {
+		let blocks = headers
+			.iter()
+			.filter_map(|hash| self.client.block_header(BlockId::Hash(*hash)))
+			.map(|header| (header, Default::default()))
+			.collect();
+
+		self.notify(blocks);
+	}
+}
+
+impl<C: BlockChainClient> ChainNotify for ChainNotificationHandler<C> {
+	fn new_blocks(
+		&self,
+		_imported: Vec<H256>,
+		_invalid: Vec<H256>,
+		enacted: Vec<H256>,
+		_retracted: Vec<H256>,
+		_sealed: Vec<H256>,
+		// Block bytes.
+		_proposed: Vec<Bytes>,
+		_duration: u64,
+	) {
+		const EXTRA_INFO_PROOF: &'static str = "Object exists in in blockchain (fetched earlier), extra_info is always available if object exists; qed";
+		let blocks = enacted
+			.into_iter()
+			.filter_map(|hash| self.client.block_header(BlockId::Hash(hash)))
+			.map(|header| {
+				let hash = header.hash();
+				(header, self.client.block_extra_info(BlockId::Hash(hash)).expect(EXTRA_INFO_PROOF))
```

Comment:
That's an implementation for full client, light implementation doesn't have `extra_info` at all.